Use metadata API in the proxy and tap injectors

alpeb opened the pull request:
Part of #9485

This adds a new `MetadataAPI` similar to the current `k8s.API` hosting informers, but backed by k8s' `metadatainformer` shared informers, which retrieves only the objects metadata, resulting in less memory consumption by its clients. Currently this is only implemented for the proxy and tap injectors. Usage by the destination controller will be implemented as a follow-up.

## Existing API enhancements

Shared objects and logic required by API and MetadataAPI have been moved to the new `k8s.go`, `api_resource.go` and `prometheus.go` files. That includes the `isValidRSParent()` function whose arg is now more generic.

## Unit tests

`/controller/k8s/api_test.go` now also instantiates a MetadataAPI, used in the augmented `TestGetObjects()` and `TestGetOwnerKindAndName()` tests. The `resources` struct was introduced to capture the common fields among tests and simplify `newMockAPI()`'s signature.

## Other Changes

The injector no longer watches for Pods. It only requires watching workloads that own resources (and also watch namespaces), so Pod is not required.

## Testing Memory Consumption

Install linkerd, inject emojivoto and check the injector memory consumption with `kubectl -n linkerd top pod linkerd-proxy-injector-xxx`. It'll start consuming about 16Mi. Then ramp up emojivoto's `voting` deployment replicas to 2000. After 5 minutes memory will stabilize around 32Mi using the current branch. Using the latest edge, it'll stabilize around 110Mi.

olix0r commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2738,8 +2738,9 @@ func CheckCanPerformAction(ctx context.Context, api *k8s.KubernetesAPI, verb, na
 // component =>
-//   pod name =>
-//     container statuses
+//
+//	pod name =>
+//	  container statuses
```
this seems unintentional

olix0r commented on pkg/k8s/k8s.go:
```diff
@@ -201,8 +206,9 @@ func ShortNameFromCanonicalResourceName(canonicalName string) string {
 // For example:
-//   `pod` -> `pod`
-//   `job` -> `k8s_job`
+//
+//	`pod` -> `pod`
+//	`job` -> `k8s_job`
```
similarly

olix0r commented on controller/k8s/api_resource.go:
```diff
@@ -0,0 +1,139 @@
+	}
+}
+
+func strToGVK(str string) schema.GroupVersionKind {
+	parts := strings.SplitN(str, ".", 3)
```
Is this called from anywhere else? If not, it seems like it might be simpler to build the structs directly instead of relying on string parsing?

alpeb replied:
Not called from anywhere else. My thinking was to make the GKV function simpler, but you're right, the added indirection isn't a brilliant idea.

mateiidavid commented on controller/k8s/api_resource.go:
```diff
@@ -0,0 +1,207 @@
+		return schema.GroupVersionKind{
+			Group:   "",
+			Version: "v1",
+			Kind:    "Configmap",
```
To be clear, unsure if this matters when the objects get deserialized/typecast. Is it possible for something not to be cast to an actual implementation if we construct an object reference with a kind that doesn't follow strict upper camelcasing?

alpeb replied:
From my testing, casing doesn't matter. And when checking debug-level logs in the injector you'll see the API calls to build the informers use lower-cased kinds.
The other place we're consuming these structs is in `GetByNamespaceFilteredCached()`, used by the injector's `webhook.go` for building the prometheus labels, which are also lower-cased.
In any case, I agree it's better to match how k8s exposes these kinds, so I'll commit all your suggestions :100:

alpeb commented on controller/k8s/api_resource.go:
```diff
@@ -0,0 +1,207 @@
+func (res APIResource) GVK() (schema.GroupVersionKind, error) {
+	switch res {
+	case CJ:
+		return schema.GroupVersionKind{
```
Good find! I'll try to add that, and chain with `.WithKind()` to generate the final GVK struct :+1: :+1:

alpeb commented on controller/k8s/api_resource.go:
```diff
@@ -0,0 +1,207 @@
+		return schema.GroupVersionKind{
+			Group:   "admissionregistration.k8s.io",
+			Version: "v1",
+			Kind:    "Mutatingwebhookconfiguration",
```
```suggestion
			Kind:    "MutatingWebhookConfiguration",
```

alpeb commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,257 @@
+) (*MetadataAPI, error) {
+	sharedInformers := metadatainformer.NewFilteredSharedInformerFactory(
+		metadataClient,
+		10*time.Minute,
```
Sounds good, I'll move that value (along with its other appearances in `api.go`) into a const in `k8s.go` :+1:

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// GetCached returns the metadata for the supplied object type and name
+func (api *MetadataAPI) GetCached(res APIResource, name string) (*metav1.PartialObjectMetadata, error) {
```
Can you clarify the naming and semantics of this?  My understanding is that this uses the metadata shared informer which should work just like the regular shared informers except that it's metadata only.  Unless there's more to it, I think calling this "cached" in the function name confuses more than it helps.

alpeb replied:
Makes sense, I can't recall why I chose that naming :man_shrugging:

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+		return nil, err
+	}
+
+	nsMeta, ok := obj.(*metav1.PartialObjectMetadata)
```
This confused me for a bit, but I understand now.  Because the metadata informers need to satisfy the informer interface, their return types are `Object` even though they will always return `*PartialObjectMetadata`.  We do this cast to return the more specific type and this cast should always succeed.  This might be worth a comment.

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+	return nsMeta, nil
+}
+
+func (api *MetadataAPI) getByNamespaceCached(res APIResource, ns, name string) (*metav1.PartialObjectMetadata, error) {
```
again, the `cached` naming is confusing, I think.

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+	return inf.Lister(), nil
+}
+
+// GetCached returns the metadata for the supplied object type and name
```
the godoc is a good place to talk about the semantics, specifically that this uses a shared informer and the results may be out of date if the informer is lagging behind.

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+	ns string,
+	name string,
+	label labels.Selector,
+) ([]*corev1.ObjectReference, error) {
```
why does this return `ObejectReference`s while the other functions return `PartialObjectMetadata`?  Should we be consistent?

alpeb replied:
This was so because the objects returned by the metadata lister didn't have their `TypeMeta` properly filled, which was a problem when issuing an `Event` in `webhook.go`. I'll push a modification to take care of that missing info so we can return `PartialObjectMetadata` here as well :+1:

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+		parentObj, err = api.getByNamespaceCached(Job, pod.Namespace, parent.Name)
+		if err != nil {
+			log.Warnf("failed to retrieve job from indexer %s/%s: %s", pod.Namespace, parent.Name, err)
+			parentObj, err = api.Client.BatchV1().Jobs(pod.Namespace).Get(ctx, parent.Name, metav1.GetOptions{})
```
here we fall back to the full k8s client, but if `api.Client` was a metadata-only client, I think that would still work for this case.

alpeb replied:
Yep, there are some small hoops to go through but we can indeed avoid the full client here :100:

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+	case "ReplicaSet":
+		var rsObj metav1.Object
+		rsObj, err = api.getByNamespaceCached(RS, pod.Namespace, parent.Name)
+		isNil := rsObj.(*metav1.PartialObjectMetadata) == nil
```
I think we can simplify this if `rsObj` is declared as a `PartialObjectMetadata` in the first place.

adleong commented on controller/k8s/metadata_api.go:
```diff
@@ -0,0 +1,256 @@
+		isNil := rsObj.(*metav1.PartialObjectMetadata) == nil
+		if err != nil {
+			log.Warnf("failed to retrieve replicaset from indexer %s/%s: %s", pod.Namespace, parent.Name, err)
+			rsObj, err = api.Client.AppsV1().ReplicaSets(pod.Namespace).Get(ctx, parent.Name, metav1.GetOptions{})
```
If `rsObj` is a `PartialObjectMetadata` and we use the metadata client as I suggest above, we'd need to add a cast here to cast the returned value to a `PartialObjectMetadata`.

alpeb replied:
After that refactoring, the metadata API will return `PartialObjectMetadata` so we're good :+1:

alpeb commented:
Thank y'all for the detailed review :pray: 

Here's the list of changes from the latest commits, addressing the feedback:

- Removed full k8s client from `MetadataAPI`, and replaced it with a reference to the metadataAPI.
- The `GetOwnerKindAndName()` method now uses that metadataAPI client to query for resources when the metadata informer doesn't return anything.
- Also added a `retry` parameter to `GetOwnerKindAndName()` like we have in the full API's `GetOwnerKindAndName()` method, to allow testing the direct API calls from the unit tests.
- Fixed the `GetByNamespaceFiltered()` method, that was ignoring the `name` parameter.
- The webhook launcher now builds the full k8s API to pass it to `NewServer()`, which uses it to build `eventBroadcaster`
- `api_test.go` and `test_helper.go` have been simplified by building the `*metav1.PartialObjectMetadata` objects automatically from the k8s manifest strings.

mateiidavid reviewed (approved):
Looks great!

alpeb commented on controller/k8s/api.go:
```diff
@@ -218,81 +190,81 @@ func newAPI(
 			api.cj = sharedInformers.Batch().V1().CronJobs()
 			api.syncChecks = append(api.syncChecks, api.cj.Informer().HasSynced)
-			api.addInformerSizeGauge("cron_job", api.cj.Informer())
+			api.promGauges.addInformerSize(k8s.CronJob, api.cj.Informer())
```
Explicitly calling through the embedded struct is used for readability; initially suggested by @adleong [above](https://github.com/linkerd/linkerd2/pull/9650#discussion_r1014292432)
